pkg/ovs/openflow/ofctrl_group.go

```diff
@@ -80,9 +83,34 @@ func (g *ofGroup) GetBundleMessage(entryOper OFOperation) (ofctrl.OpenFlowModMes
 		operation = openflow15.OFPGC_MODIFY
 	case DeleteMessage:
 		operation = openflow15.OFPGC_DELETE
+		// If the operation is to delete the group, empty the slice storing buckets since the number of buckets could
+		// be greater than MaxBucketsPerMessage.
+		g.ofctrl.Buckets = nil
+	}
+
+	var messages []ofctrl.OpenFlowModMessage
+	for i, done := 0, false; done != true; i++ {
+		// Get the range of buckets to generate a temp group.
+		firstBucketIdx := i * MaxBucketsPerMessage
+		lastBucketIdx := (i + 1) * MaxBucketsPerMessage
+		if lastBucketIdx > len(g.ofctrl.Buckets) {
+			lastBucketIdx = len(g.ofctrl.Buckets)
+			done = true
+		}
+		// Generate a temp group to get an OVS message. Note that, the original group should not be modified since it is
+		// also stored in group cache, and the group cache is used when replying groups.
+		groupMessage := &ofctrl.Group{
+			ID:        g.ofctrl.ID,
+			GroupType: g.ofctrl.GroupType,
+			Buckets:   g.ofctrl.Buckets[firstBucketIdx:lastBucketIdx],
+		}
```

Comment:
what if number of buckets equals MaxBucketsPerMessage? Is it legal to insert empty buckets?

I think the code could be simplified:
```
	for start := 0; start < len(g.ofctrl.Buckets); start += MaxBucketsPerMessage {
		// Get the range of buckets to generate a temp group.
		end := start + MaxBucketsPerMessage
		if end > len(g.ofctrl.Buckets) {
			end = len(g.ofctrl.Buckets)
		}
		// Generate a temp group to get an OVS message. Note that, the original group should not be modified since it is
		// also stored in group cache, and the group cache is used when replying groups.
		groupMessage := &ofctrl.Group{
			ID:        g.ofctrl.ID,
			GroupType: g.ofctrl.GroupType,
			Buckets:   g.ofctrl.Buckets[start:end],
		}
		// For the message which is not the first, insert_buckets is used to add buckets to the group on OVS.
		if start != 0 {
			operation = openflow15.OFPGC_INSERT_BUCKET
		}
		messages = append(messages, groupMessage.GetBundleMessage(operation))
	}
```

Reply:
Looks better, and I changed a little to support a message when there is no bucket in the group.